design/design-doc-composition-functions.md

```diff
@@ -175,32 +200,606 @@ metadata:
+design][alpha-design] with the [`FunctionIO`][functionio-schema] elevated from
+opaque YAML-encoded bytes to 'native' RPC code. Kubernetes resources are
+represented using the [`google.protobuf.Struct` well-known
+type][google-protobuf-struct], which can encode arbitrary JSON.
+
+Some key differences between the alpha `FunctionIO` and the proposed beta
+`RunFunctionRequest`:
+
+* `observed.resources` and `desired.resources` are a map keyed by resource name,
+  not an array of objects with name fields. The previous pattern was a result of
+  `FunctionIO` attempting to be an idiomatic KRM object, which is not a
+  constraint for `RunFunctionRequest`. This is should make it easier to lookup
+  resources by name when developing Functions.
+* Entries in `desired.resources` can no longer return 'derived' connection
+  secrets or readiness checks, similar to an entry in the P&T resources array.
+  Instead a Function is intended to set XR connection details and/or readiness
+  by mutating them directly.
+
+The package manager is responsible for creating a headless Kubernetes Service
+where each Function's Deployment can be reached. The address of the Service will
+be exposed as the `status.endpoint` of the Function resource. The Service must
+be headless in order for Crossplane's gRPC client to load-balance connections
+when there are multiple Function replicas.
+
+Note that the fact this endpoint is powered by a Service is an implementation
+detail; it may be possible for Functions to be reached (and indeed deployed) by
+other means in future (see [Runtime Configuration](#runtime-configuration)).
+
+```yaml
+apiVersion: pkg.crossplane.io/v1beta1
+kind: Function
+metadata:
+  name: go-templates
+spec:
+  package: xpkg.upbound.io/negz/go-templates:v0.1.0
+status:
+  # The gRPC endpoint where Crossplane will send RunFunctionRequests.
+  endpoint: https://go-templates-9sdfn2
+```
+
+gRPC communication between Crossplane and a Function will be secured by mutual
+transport layer security (mTLS). Crossplane has an established pattern for
+setting up mTLS certificate bundles - we do this today to secure communication
+between the API server and our validation webhooks, and between Crossplane and
+External Secret Stores. We might want to [rethink this][cert-per-entity] in
+future, but to begin with we'll stick with the established pattern, i.e.:
+
+* The Helm chart creates an empty Secret for the client bundle.
+* The core init container bootstraps the client bundle.
+* The package manager creates a Secret for each Function's server bundle.
+* The package manager bootstraps each Function's server bundle.
+
+See [#3884] for an example of this pattern.
+
+The frequency with which Crossplane will call each Function is relative to how
+frequently each XR is reconciled. There are two main triggers to reconcile an XR:
+
+1. A watch-triggered reconcile, i.e. because the XR changed.
+2. A poll-triggered reconcile, by default once per 60 seconds.
```

Comment:
no watch on the composition?

Reply:
There is not today iirc. I'm not sure we'd need one - you need to update the XR to use a new CompositionRevision iirc. So a change of Composition would require changing the XR. That said, I could be forgetting something. No concerns about watching Composition (or CompositionRevision) if it helps.

Reply:
Added a section making this explicit.

Reply:
just thinking that this matters in the development workflow. In prod, compositions never change, in dev's inner loop all the time. So a watch would be nice.